bytecode/src/function/instructions/add_wrapped.rs

```diff
@@ -0,0 +1,178 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    function::{parsers::*, Instruction, Opcode, Operation, Registers},
+    helpers::Register,
+    Program,
+    Value,
+};
+use snarkvm_circuits::{AddWrapped as AddWrappedCircuit, Literal, Parser, ParserResult};
+use snarkvm_utilities::{FromBytes, ToBytes};
+
+use core::fmt;
+use nom::combinator::map;
+use std::io::{Read, Result as IoResult, Write};
+
+/// Adds `first` with `second`, wrapping around on overflow, and storing the outcome in `destination`.
+pub struct AddWrapped<P: Program> {
+    operation: BinaryOperation<P>,
+}
+
+impl<P: Program> AddWrapped<P> {
+    /// Returns the operands of the instruction.
+    pub fn operands(&self) -> Vec<Operand<P>> {
+        self.operation.operands()
+    }
+
+    /// Returns the destination register of the instruction.
+    pub fn destination(&self) -> &Register<P> {
+        self.operation.destination()
+    }
+}
+
+impl<P: Program> Opcode for AddWrapped<P> {
+    /// Returns the opcode as a string.
+    #[inline]
+    fn opcode() -> &'static str {
+        "add.w"
+    }
+}
+
+impl<P: Program> Operation<P> for AddWrapped<P> {
+    /// Evaluates the operation.
+    #[inline]
+    fn evaluate(&self, registers: &Registers<P>) {
+        // Load the values for the first and second operands.
+        let first = match registers.load(self.operation.first()) {
+            Value::Literal(literal) => literal,
+            Value::Composite(name, ..) => P::halt(format!("{name} is not a literal")),
+        };
+        let second = match registers.load(self.operation.second()) {
+            Value::Literal(literal) => literal,
+            Value::Composite(name, ..) => P::halt(format!("{name} is not a literal")),
+        };
+
+        // Perform the operation.
+        let result = match (first, second) {
+            (Literal::I8(a), Literal::I8(b)) => Literal::I8(a.add_wrapped(&b)),
+            (Literal::I16(a), Literal::I16(b)) => Literal::I16(a.add_wrapped(&b)),
+            (Literal::I32(a), Literal::I32(b)) => Literal::I32(a.add_wrapped(&b)),
+            (Literal::I64(a), Literal::I64(b)) => Literal::I64(a.add_wrapped(&b)),
+            (Literal::I128(a), Literal::I128(b)) => Literal::I128(a.add_wrapped(&b)),
+            (Literal::U8(a), Literal::U8(b)) => Literal::U8(a.add_wrapped(&b)),
+            (Literal::U16(a), Literal::U16(b)) => Literal::U16(a.add_wrapped(&b)),
+            (Literal::U32(a), Literal::U32(b)) => Literal::U32(a.add_wrapped(&b)),
+            (Literal::U64(a), Literal::U64(b)) => Literal::U64(a.add_wrapped(&b)),
+            (Literal::U128(a), Literal::U128(b)) => Literal::U128(a.add_wrapped(&b)),
+            _ => P::halt(format!("Invalid '{}' instruction", Self::opcode())),
+        };
```

Comment:
a macro could shorten `match` expressions like this; I haven't tested it, but it would look something like
```
macro_rules! impl_evaluate {
    ($($ty:ident),*) => {
        #[inline]
        fn evaluate(&self, registers: &Registers<P>) {
            (...)

            let result = match (first, second) {
                $( (Literal::$ty(a), Literal::$ty(b)) => Literal::$ty(a.add_wrapped(&b)), )*
                _ => P::halt(format!("Invalid '{}' instruction", Self::opcode())),
            };
        }
    }
}
```
and called like
```
impl<P: Program> Operation<P> for AddWrapped<P> {
    impl_evaluate!(I8, I16, I32, I64, I128, U8, U16, U32, U64, U128);
}
```
it might not apply to all implementations of `Operation`, but it's there's possibly some room for code deduplication

Reply:
I think in this case I'd prefer to keep this simple and readable, even if there is code duplication.